<div style="display:block" direction=rtl align=right><br><br>
    <div  style="width:100%;margin:100;display:block"  display=block align=center>
        <img width=130 align=right src="https://i.ibb.co/yXKQmtZ/logo1.png" style="margin:0;" />
        <img width=170 align=left  src="https://i.ibb.co/wLjqFkw/logo2.png" style="margin:0;" />
        <span><br><font size=5>University of Tehran , Iran's Artificial Intelligence House</font></span>
        <span><br><font size=3>Data Science Bootcamp</font></span>
        <span><br><font size=3>Spring 2025</font></span>
    </div><br><br><br>
    <div style="display:block" align=left display=block>
        <font size=3>Work with Python</font><br>
        <hr />
        <font size=3>Developer: <a href="mailto:p.zilouchian@ut.ac.ir">Parham Zilouchian Moghaddam</a><br></font><br>
    </div>
</div>

<a id='work_with_missing_data'></a>
# Work with missing data in a Python notebook

Throughout the following exercises, you will be discovering and working with missing data on a dataset.  Before starting on this programming exercise, we strongly recommend watching the video lecture and completing the IVQ for the associated topics.

All the information you need for solving this assignment is in this notebook, and all the code you will be implementing will take place within this notebook.

As we move forward, you can find instructions on how to install required libraries as they arise in this notebook. Before we begin with the exercises and analyzing the data, we need to import all libraries and extensions required for this programming exercise. Throughout the course, we will be using pandas, numpy, datetime, for operations, and matplotlib, pyplot and seaborn for plotting.

## Objective

We will be examining lightning strike data collected by the National Oceanic and Atmospheric Association (NOAA) for the month of August 2018. There are two datasets. The first includes five columns:  

|date|center_point_geom|longitude|latitude|number_of_strikes|
|---|---|---|---|---|

The second dataset contains seven columns:

|date|zip_code|city|state|state_code|center_point_geom|number_of_strikes|
|---|---|---|---|---|---|---|  

The first dataset has two unique colums: `longitude` and `latitude`.  
The second dataset has four unique columns: `zip_code`, `city`, `state`, and `state_code`.  
There are three columns that are common between them: `date`, `center_point_geom`, and `number_of_strikes`.

We want to combine the two datasets into a single dataframe that has all of the information from both datasets. Ideally, both datasets will have the same number of entries for the same locations on the same dates. If they don't, we'll investigate which data is missing.

In [1]:
# Import statements
import pandas as pd
import numpy as np
import seaborn as sns
import datetime
from matplotlib import pyplot as plt

In [2]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# Read in first dataset
df = pd.read_csv('/content/drive/MyDrive/eda_missing_data_dataset1.csv')

In [4]:
# Print the first 5 rows of dataset 1
df.head()

date  center_point_geom  longitude  latitude  number_of_strikes
0  2018-08-01  POINT(-81.6 22.6)      -81.6      22.6                 48
1  2018-08-01  POINT(-81.1 22.6)      -81.1      22.6                 32
2  2018-08-01  POINT(-80.9 22.6)      -80.9      22.6                118
3  2018-08-01  POINT(-80.8 22.6)      -80.8      22.6                 69
4  2018-08-01  POINT(-98.4 22.8)      -98.4      22.8                 44

Let's check on our dataset shape to determine number of columns and rows.

In [5]:
df.shape

(717530, 5)

Now we'll read in the second dataset.

In [6]:
# Read in second dataset
df_zip = pd.read_csv('/content/drive/MyDrive/eda_missing_data_dataset2.csv')

In [7]:
# Print the first 5 rows of dataset 2
df_zip.head()

date  zip_code                                 city          state  \
0  2018-08-08      3281                                Weare  New Hampshire   
1  2018-08-14      6488                 Heritage Village CDP    Connecticut   
2  2018-08-16     97759  Sisters city, Black Butte Ranch CDP         Oregon   
3  2018-08-18      6776                      New Milford CDP    Connecticut   
4  2018-08-08      1077                            Southwick  Massachusetts   

  state_code   center_point_geom  number_of_strikes  
0         NH   POINT(-71.7 43.1)                  1  
1         CT   POINT(-73.2 41.5)                  3  
2         OR  POINT(-121.4 44.3)                  3  
3         CT   POINT(-73.4 41.6)                 48  
4         MA     POINT(-72.8 42)                  2

And check the shape...

In [8]:
df_zip.shape

(323700, 7)

Hmmm... This dataset has less than half the number of rows as the first one. But which ones are they?  

The first thing we'll do to explore this discrepancy is join the two datasets into a single dataframe. We can do this using the `merge()` method of the `DataFrame` class. For more information about the `merge()` method, refer to the [merge() pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html).  

Begin with the first dataframe (`df`) and call the `merge()` method on it. The first argument is a positional argument that specifies the dataframe we want to merge with, known as the `right` dataframe. (The dataframe you're calling the method on is always the `left` dataframe.) The `how` argument specifies which dataframe's keys we'll use to match to, and the `on` argument lets us define the columns to use as keys.

A demonstration will make this easiest to understand. Refer to the **[BONUS CONTENT](#merge_bonus)** at the end of the notebook for different examples of the `merge()` method.
  

In [9]:
# Left-join the two datasets
df_joined = df.merge(df_zip, how='left', on=['date','center_point_geom'])

In [10]:
# Print the first 5 rows of the merged data
df_joined.head()

date  center_point_geom  longitude  latitude  number_of_strikes_x  \
0  2018-08-01  POINT(-81.6 22.6)      -81.6      22.6                   48   
1  2018-08-01  POINT(-81.1 22.6)      -81.1      22.6                   32   
2  2018-08-01  POINT(-80.9 22.6)      -80.9      22.6                  118   
3  2018-08-01  POINT(-80.8 22.6)      -80.8      22.6                   69   
4  2018-08-01  POINT(-98.4 22.8)      -98.4      22.8                   44   

   zip_code city state state_code  number_of_strikes_y  
0       NaN  NaN   NaN        NaN                  NaN  
1       NaN  NaN   NaN        NaN                  NaN  
2       NaN  NaN   NaN        NaN                  NaN  
3       NaN  NaN   NaN        NaN                  NaN  
4       NaN  NaN   NaN        NaN                  NaN

Notice that the new dataframe has all of the columns of both original dataframes, and it has two `number_of_strikes` columns that are suffixed with `_x` and `_y`. This is because the key columns from both dataframes were the same, so they appear once in the merged dataframe. The unique columns of each original dataframe also appear in the merged dataframe. But both original dataframes had another column&mdash;`number_of_strikes`&mdash;that had the same name in both dataframes and was not indicated as a key. Pandas handles this by adding both columns to the new dataframe.

Now we'll check the summary on this joined dataset.  

In [11]:
# Get descriptive statistics of the joined dataframe
df_joined.describe()

longitude       latitude  number_of_strikes_x       zip_code  \
count  717530.000000  717530.000000        717530.000000  323700.000000   
mean      -90.875445      33.328572            21.637081   57931.958996   
std        13.648429       7.938831            48.029525   22277.327411   
min      -133.900000      16.600000             1.000000    1002.000000   
25%      -102.800000      26.900000             3.000000   38260.750000   
50%       -90.300000      33.200000             6.000000   59212.500000   
75%       -80.900000      39.400000            21.000000   78642.000000   
max       -43.800000      51.700000          2211.000000   99402.000000   

       number_of_strikes_y  
count        323700.000000  
mean             25.410587  
std              57.421824  
min               1.000000  
25%               3.000000  
50%               8.000000  
75%              24.000000  
max            2211.000000

The count information confirms that the new dataframe is missing some data.

Now let's check how many missing state locations we have by using `isnull()` to create a Boolean mask that we'll apply to `df_joined`. The mask is a pandas `Series` object that contains `True` for every row with a missing `state_code` value and `False` for every row that is not missing data in this column. When the mask is applied to `df_joined`, it filters out the rows that are not missing `state_code` data. (Note that using the `state_code` column to create this mask is an arbitrary decision. We could have selected `zip_code`, `city`, or `state` instead and gotten the same results.)

In [12]:
# Create a new df of just the rows that are missing data
df_null_geo = df_joined[pd.isnull(df_joined.state_code)]
df_null_geo.shape

(393830, 10)

We can confirm that `df_null_geo` contains only the rows with the missing `state_code` values by using the `info()` method on `df_joined` and comparing.

In [13]:
# Get non-null counts on merged dataframe
df_joined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717530 entries, 0 to 717529
Data columns (total 10 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   date                 717530 non-null  object 
 1   center_point_geom    717530 non-null  object 
 2   longitude            717530 non-null  float64
 3   latitude             717530 non-null  float64
 4   number_of_strikes_x  717530 non-null  int64  
 5   zip_code             323700 non-null  float64
 6   city                 323700 non-null  object 
 7   state                323700 non-null  object 
 8   state_code           323700 non-null  object 
 9   number_of_strikes_y  323700 non-null  float64
dtypes: float64(4), int64(1), object(5)
memory usage: 54.7+ MB


If we subtract the 323,700 non-null rows in columns 5-9 of `df_joined` from the 717,530 non-null rows in columns 0-4 of `df_joined`, we're left with 393,830 rows that contain missing data&mdash;the same number of rows contained in `df_null_geo`.

In [14]:
# Print the first 5 rows
df_null_geo.head()

date  center_point_geom  longitude  latitude  number_of_strikes_x  \
0  2018-08-01  POINT(-81.6 22.6)      -81.6      22.6                   48   
1  2018-08-01  POINT(-81.1 22.6)      -81.1      22.6                   32   
2  2018-08-01  POINT(-80.9 22.6)      -80.9      22.6                  118   
3  2018-08-01  POINT(-80.8 22.6)      -80.8      22.6                   69   
4  2018-08-01  POINT(-98.4 22.8)      -98.4      22.8                   44   

   zip_code city state state_code  number_of_strikes_y  
0       NaN  NaN   NaN        NaN                  NaN  
1       NaN  NaN   NaN        NaN                  NaN  
2       NaN  NaN   NaN        NaN                  NaN  
3       NaN  NaN   NaN        NaN                  NaN  
4       NaN  NaN   NaN        NaN                  NaN

Now that we've merged all of our data together and isolated the rows with missing data, we can better understand what data is missing by plotting the longitude and latitude of locations that are missing city, state, and zip code data.

In [15]:
# Create new df of just latitude, longitude, and number of strikes and group by latitude and longitude
top_missing = df_null_geo[['latitude','longitude','number_of_strikes_x']
            ].groupby(['latitude','longitude']
                      ).sum().sort_values('number_of_strikes_x',ascending=False).reset_index()
top_missing.head(10)

latitude  longitude  number_of_strikes_x
0      22.4      -84.2                 3841
1      22.9      -82.9                 3184
2      22.4      -84.3                 2999
3      22.9      -83.0                 2754
4      22.5      -84.1                 2746
5      22.5      -84.2                 2738
6      22.3      -81.0                 2680
7      22.9      -82.4                 2652
8      22.9      -82.3                 2618
9      22.3      -84.3                 2551

Let's import plotly to reduce the size of the data frame as we create a geographic scatter plot.

In [16]:
import plotly.express as px  # Be sure to import express
# reduce size of db otherwise it could break
fig = px.scatter_geo(top_missing[top_missing.number_of_strikes_x>=300],  # Input Pandas DataFrame
                    lat="latitude",  # DataFrame column with latitude
                    lon="longitude",  # DataFrame column with latitude
                    size="number_of_strikes_x") # Set to plot size as number of strikes
fig.update_layout(
    title_text = 'Missing data', # Create a Title
)

fig.show()

It’s a nice geographic visualization, but we really don’t need the global scale. Let’s scale it down to only the geographic area we are interested in - the United States.

**Note:** The following cell's output is viewable in two ways: You can re-run this cell (and all of the ones before it) or manually convert the notebook to "Trusted."

In [17]:
import plotly.express as px  # Be sure to import express
fig = px.scatter_geo(top_missing[top_missing.number_of_strikes_x>=300],  # Input Pandas DataFrame
                    lat="latitude",  # DataFrame column with latitude
                    lon="longitude",  # DataFrame column with latitude
                    size="number_of_strikes_x") # Set to plot size as number of strikes
fig.update_layout(
    title_text = 'Missing data', # Create a Title
    geo_scope='usa',  # Plot only the USA instead of globe
)

fig.show()

This explains why so many rows were missing state and zip code data! Most of these lightning strikes occurred over water&mdash;the Atlantic Ocean, the Sea of Cortez, the Gulf of Mexico, the Caribbean Sea, and the Great Lakes. Of the strikes that occurred over land, most of those were in Mexico, the Bahamas, and Cuba&mdash;places outside of the U.S. and without U.S. zip codes. Nonetheless, some of the missing data is from Florida and elsewhere within the United States, and we might want to ask the database owner about this.

If you have successfully completed the material above, congratulations! You now understand handling missing data in Python and should be able to start using it on your own datasets.

<a id='merge_bonus'></a>
### Bonus: `df.merge()` demonstration:

Begin with two dataframes:

In [18]:
# Define df1
data = {'planet': ['Mercury', 'Venus', 'Earth', 'Mars',
                    'Jupiter', 'Saturn', 'Uranus', 'Neptune'],
        'radius_km': [2440, 6052, 6371, 3390, 69911, 58232,
                      25362, 24622],
        'moons': [0, 0, 1, 2, 80, 83, 27, 14]
         }
df1 = pd.DataFrame(data)
df1

planet  radius_km  moons
0  Mercury       2440      0
1    Venus       6052      0
2    Earth       6371      1
3     Mars       3390      2
4  Jupiter      69911     80
5   Saturn      58232     83
6   Uranus      25362     27
7  Neptune      24622     14

In [19]:
# Define df2
data = {'planet': ['Mercury', 'Venus', 'Earth', 'Meztli', 'Janssen'],
        'radius_km': [2440, 6052, 6371, 48654, 11959],
        'life?': ['no', 'no', 'yes', 'no', 'yes'],
         }
df2 = pd.DataFrame(data)
df2

planet  radius_km life?
0  Mercury       2440    no
1    Venus       6052    no
2    Earth       6371   yes
3   Meztli      48654    no
4  Janssen      11959   yes

Now we'll merge the two dataframes on the `['planet', 'radius_km']` columns. Try running the below cell with each of the following arguments for the **`how`** keyword: `'left'`, `'right'`, `'inner'`, and `'outer'`. Notice how each argument changes the result.  

Feel free to change the columns specified by the **`on`** argument too!

In [20]:
merged = df1.merge(df2, how='left', on=['planet', 'radius_km'])
merged

planet  radius_km  moons life?
0  Mercury       2440      0    no
1    Venus       6052      0    no
2    Earth       6371      1   yes
3     Mars       3390      2   NaN
4  Jupiter      69911     80   NaN
5   Saturn      58232     83   NaN
6   Uranus      25362     27   NaN
7  Neptune      24622     14   NaN